File: frontend/Pseudo2code+/src/services/tasks/taskResolver.js
import { TASK_DEFINITIONS } from './taskDefinitions'

export function resolveSuggestedTasks(user, limit = 4) {
  if (!user) return []

  return TASK_DEFINITIONS
    .filter(task => {
      // ⛔ once task already completed by data
      if (task.once && typeof task.condition === "function") {
        return task.condition(user)
      }

      // normal condition
      if (typeof task.condition === "function") {
        return task.condition(user)
      }

      return true
    })
    .sort((a, b) => {
      const pa = typeof a.priority === "function" ? a.priority(user) : 0
      const pb = typeof b.priority === "function" ? b.priority(user) : 0
      return pb - pa
    })
    .slice(0, limit)
}

File: backend/services/tasks/task_resolver.py
import random
from services.tasks.task_definitions import get_task_definitions
from services.tasks.task_engine import is_task_completed


def resolve_suggested_tasks(user: dict, limit: int = 4):
    tasks = get_task_definitions()
    completed_ids = {
        t["task_id"] for t in user.get("user_tasks", {}).get("completed", [])
    }

    eligible = []

    for task in tasks:
        if task.get("once") and task["id"] in completed_ids:
            continue

        if not task["condition"](user):
            continue

        eligible.append(task)

    # 🔥 sort safely
    eligible.sort(key=lambda t: t.get("priority", 0), reverse=True)

    # 🔄 rotate top candidates
    top = eligible[:6]
    random.shuffle(top)
    eligible[:6] = top

    return eligible[:limit]
